test/gleam/regex_test.gleam

```diff
@@ -84,4 +84,21 @@ pub fn scan_test() {
 
   regex.scan(re, "你好 42 世界")
   |> should.equal([Match(content: "42", submatches: [Some("42")])])
+
+  let assert Ok(re) = regex.from_string("([+|\\-])?(\\d+)(\\w+)?")
+  regex.scan(re, "+36kg")
+  |> should.equal([
+    Match(content: "+36kg", submatches: [Some("+"), Some("36"), Some("kg")]),
+  ])
+
+  regex.scan(re, "36kg")
+  |> should.equal([
+    Match(content: "36kg", submatches: [None, Some("36"), Some("kg")]),
+  ])
+
+  regex.scan(re, "-36")
+  |> should.equal([Match(content: "-36", submatches: [Some("-"), Some("36")])])
```

Comment:
Sorry, I'm a bit confused by this one. Why is it that the first case returns `None` when the group doesn't capture, yet the second one is absent?

Wouldn't having the number of captures be variable make it harder to know which value is for which group?

Reply:
unfortunately that's not something we can control.
javascript returns `undefined` when a capture group does not capture anything, which would solve the problem that you mentioned.

But erlang does not return anything and that's exactly the source of the inconsistency.
You either need to add the missing ones in the erlang side, or remove the `undefined` ones from the javascript side.
But to add the missing ones in the erlang side, you would need to parse the regex yourself and figure out which capture groups didn't capture anything and I don't think we'd want to do that.(I may be wrong here, so tell me if this is actually possible) So the only choice is to remove the `undefined` ones from the javascript side.

Of course even erlang realizes that this would only work if any capture group that doesn't capture anything but is left of the capture groups that did capture something, still return something, otherwise you couldn't keep track of it at all. So if "36kg" would give "[Some("36"), Some("kg")]", you couldn't figure out that the first capture group didn't capture anything, that's why erlang gives "[None, Some("36"), Some("kg")]" instead.

Reply:
So the behaviour is that empty capture groups are retained unless there's no full capture groups after it, in which case it is dropped? Is that right?

Does Elixir also do this?

Reply:
Yes, that's the behavior and yes, that's how elixir does it too. It makes sense that elixir just uses the regex capabilities of erlang itself.
As long as there is going to be a non-empty capture later, any empty capture is still returned. But as soon as capture groups stop capturing anything, nothing will be returned for capture groups that didn't capture anything.

Reply:
I should also point out that if you give add `{capture, all, binary}` to the `re:run` options, the return values will become binaries and you won't need to use the `{Start, Length}` and `binary_part` to get them.